### DISCUSSION TITLE
Operator '==' cannot be applied to operands of type 'TProperty' and 'TProperty'

### DISCUSSION THREAD

**Discussion by xiaodongcool** [dotnet/csharplang#6445] [Q&A] [answered]
```
public interface IFoo<TProperty>
{
    public TProperty Name { get; set; }
}

public class Program<TClass, TProperty> where TClass : IFoo<TProperty>
{
    public void Main(TProperty name)
    {
        // error
        TestExpression(_ => _.Name == name);
    }

    public void TestExpression(Expression<Func<TClass, bool>>? match = null) { }
}
```

**Comment by FaustVX** [comment, marked as answer]
Because the compiler can't know if `TProperty` declare an `==` operator.
You can do `_ => _.Name.Equals(name)`

**Reply by tfenise** [reply]
or `_ => System.Collections.Generic.EqualityComparer<TProperty>.Default.Equals(_.Name, name)`, which is generic and may eliminate boxing for value types.

**Reply by Joe4evr** [reply]
It's not just that it'd remove boxing, but if the JIT sees the full expression of `EqualityComparer<T>.Default.Equals(x, y)` it will simply emit the same asm as it would have for `x == y`.
